From 87ab717f54fb2e18e53332813e7f5c3b64c2011c Mon Sep 17 00:00:00 2001
From: Mandy Lavi <mandy.lavi@freescale.com>
Date: Sun, 23 Feb 2014 16:06:09 +0200
Subject: [PATCH 0762/1089] fmd: fix mis-assignment of compat and pcd
 structures

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

Fixes memory corruption issues within lnxwrp_ioctls_fm.c
Caused by wrong usage of some data structures.

Change-Id: I75b710dd5888cc6ab3b86f0604cd5ef177396e23
Signed-off-by: Mandy Lavi <mandy.lavi@freescale.com>
Reviewed-on: http://git.am.freescale.net:8181/9041
Reviewed-by: Nir Erez <nir.erez@freescale.com>
Reviewed-by: Jose Rivera <German.Rivera@freescale.com>
Tested-by: Jose Rivera <German.Rivera@freescale.com>
Reviewed-on: http://git.am.freescale.net:8181/9386
Reviewed-by: Mandy Lavi <Mandy.Lavi@freescale.com>
---
 .../freescale/fman/src/wrapper/lnxwrp_ioctls_fm.c        | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/drivers/net/ethernet/freescale/fman/src/wrapper/lnxwrp_ioctls_fm.c b/drivers/net/ethernet/freescale/fman/src/wrapper/lnxwrp_ioctls_fm.c
index 990092e..473c3b0 100644
--- a/drivers/net/ethernet/freescale/fman/src/wrapper/lnxwrp_ioctls_fm.c
+++ b/drivers/net/ethernet/freescale/fman/src/wrapper/lnxwrp_ioctls_fm.c
@@ -1373,7 +1373,7 @@ Status: feature not supported
                     RETURN_ERROR(MINOR, E_NO_MEMORY, ("IOCTL FM PCD"));
                 }
 
-                memset(compat_param, 0, sizeof(ioc_fm_pcd_plcr_profile_params_t));
+                memset(compat_param, 0, sizeof(ioc_compat_fm_pcd_plcr_profile_params_t));
                 if (copy_from_user(compat_param, (
                             ioc_compat_fm_pcd_plcr_profile_params_t *)compat_ptr(arg),
                             sizeof(ioc_compat_fm_pcd_plcr_profile_params_t)))
@@ -2719,7 +2719,7 @@ invalid_port_id:
 		if (!param)
 			RETURN_ERROR(MINOR, E_NO_MEMORY, ("IOCTL FM PCD"));
 
-		memset(param, 0, sizeof(ioc_fm_pcd_plcr_profile_params_t));
+		memset(param, 0, sizeof(ioc_fm_pcd_frm_replic_group_params_t));
 
 #if defined(CONFIG_COMPAT)
 		if (compat)
@@ -2729,7 +2729,7 @@ invalid_port_id:
 
 			compat_param =
 				(ioc_compat_fm_pcd_frm_replic_group_params_t *)
-					XX_Malloc(sizeof(*compat_param));
+					XX_Malloc(sizeof(ioc_compat_fm_pcd_frm_replic_group_params_t));
 			if (!compat_param)
 			{
 				XX_Free(param);
@@ -2737,11 +2737,11 @@ invalid_port_id:
 						("IOCTL FM PCD"));
 			}
 
-			memset(compat_param, 0, sizeof(*compat_param));
+			memset(compat_param, 0, sizeof(ioc_compat_fm_pcd_frm_replic_group_params_t));
 			if (copy_from_user(compat_param,
 				(ioc_compat_fm_pcd_frm_replic_group_params_t *)
 					compat_ptr(arg),
-					sizeof(*compat_param))) {
+					sizeof(ioc_compat_fm_pcd_frm_replic_group_params_t))) {
 				XX_Free(compat_param);
 				XX_Free(param);
 				RETURN_ERROR(MINOR, E_READ_FAILED, NO_MSG);
@@ -2786,7 +2786,7 @@ invalid_port_id:
 
 			compat_param =
 				(ioc_compat_fm_pcd_frm_replic_group_params_t *)
-					XX_Malloc(sizeof(*compat_param));
+					XX_Malloc(sizeof(ioc_compat_fm_pcd_frm_replic_group_params_t));
 			if (!compat_param)
 			{
 				XX_Free(param);
@@ -2794,14 +2794,14 @@ invalid_port_id:
 						("IOCTL FM PCD"));
 			}
 
-			memset(compat_param, 0, sizeof(*compat_param));
+			memset(compat_param, 0, sizeof(ioc_compat_fm_pcd_frm_replic_group_params_t));
 			compat_copy_fm_pcd_frm_replic_group_params(compat_param,
 					param, COMPAT_K_TO_US);
 			if (copy_to_user(
 				(ioc_compat_fm_pcd_frm_replic_group_params_t *)
 					compat_ptr(arg),
 					compat_param,
-					sizeof(*compat_param)))
+					sizeof(ioc_compat_fm_pcd_frm_replic_group_params_t)))
 				err = E_WRITE_FAILED;
 
 			XX_Free(compat_param);
-- 
2.0.2

